File: src/core/link.ts
const SEARCH_MARKER = "#!/search/timeshift";
const DETAIL_PREFIX = "#!/ts/";

/**
 * Source map in this file:
 * - All functions are new CLI helpers, designed to parse/route Radiko hash URLs.
 */
export type RadikoLinkKind = "search" | "detail" | "unsupported";

export interface DetailRef {
  stationId: string;
  ft: string;
}

/**
 * Source: New helper for CLI URL routing (not directly extracted from rajiko).
 */
export function classifyRadikoLink(url: string): RadikoLinkKind {
  if (url.includes(SEARCH_MARKER)) {
    return "search";
  }
  if (url.includes(DETAIL_PREFIX)) {
    return "detail";
  }
  return "unsupported";
}

/**
 * Source: New helper for CLI detail URL parsing (compatible with rajiko URL format).
 */
export function extractDetailFromDetailUrl(url: string): DetailRef {
  const normalized = new URL(url);
  const hash = normalized.hash.startsWith("#!")
    ? normalized.hash.slice(2)
    : normalized.hash.slice(1);
  const segments = hash.split("/").filter(Boolean);
  if (segments.length < 3 || segments[0] !== "ts") {
    throw new Error(`Invalid detail URL: ${url}`);
  }

  const stationId = segments[1];
  const ft = segments[2];
  if (!/^\d{14}$/.test(ft)) {
    throw new Error(`Invalid ft timestamp in detail URL: ${url}`);
  }

  return { stationId, ft };
}

/**
 * Source: New helper for search-page result selection.
 */
export function pickLatestDetailUrl(
  urls: string[],
  now: Date = new Date(),
): string {
  const nowTs = toTimestamp(now);
  const details = urls
    .map((url) => ({ url, detail: extractDetailFromDetailUrl(url) }))
    .filter((x) => x.detail.ft <= nowTs)
    .sort((a, b) => b.detail.ft.localeCompare(a.detail.ft));

  if (details.length === 0) {
    throw new Error("No usable detail URL found in search results.");
  }
  return details[0].url;
}

function toTimestamp(value: Date): string {
  const y = value.getFullYear().toString().padStart(4, "0");
  const m = (value.getMonth() + 1).toString().padStart(2, "0");
  const d = value.getDate().toString().padStart(2, "0");
  const hh = value.getHours().toString().padStart(2, "0");
  const mm = value.getMinutes().toString().padStart(2, "0");
  const ss = value.getSeconds().toString().padStart(2, "0");
  return `${y}${m}${d}${hh}${mm}${ss}`;
}

File: internal/domain/link.go
package domain

import (
	"fmt"
	"net/url"
	"sort"
	"strings"
	"time"

	"rajidou/internal/util"
)

// Source map in this file:
//   - URL classification/parsing helpers are CLI-specific,
//     compatible with Radiko timeshift URL format used by Rajiko.
const (
	searchMarker = "#!/search/timeshift"
	detailPrefix = "#!/ts/"
)

// LinkKind classifies Radiko links by workflow entry point.
type LinkKind string

const (
	// LinkKindSearch identifies a search result URL that must be resolved.
	LinkKindSearch LinkKind = "search"
	// LinkKindDetail identifies a direct timeshift detail URL.
	LinkKindDetail LinkKind = "detail"
	// LinkKindUnsupported identifies links outside supported timeshift flows.
	LinkKindUnsupported LinkKind = "unsupported"
)

// DetailRef is the normalized station/time identifier extracted from detail URLs.
type DetailRef struct {
	StationID string
	FT        string
}

// ClassifyRadikoLink returns which timeshift flow should handle the URL.
func ClassifyRadikoLink(raw string) LinkKind {
	if strings.Contains(raw, searchMarker) {
		return LinkKindSearch
	}
	if strings.Contains(raw, detailPrefix) {
		return LinkKindDetail
	}
	return LinkKindUnsupported
}

// ExtractDetailFromDetailURL parses a Radiko detail URL and validates its FT
// timestamp format to avoid propagating malformed identifiers downstream.
func ExtractDetailFromDetailURL(raw string) (DetailRef, error) {
	u, err := url.Parse(raw)
	if err != nil {
		return DetailRef{}, fmt.Errorf("invalid detail URL: %s", raw)
	}
	hash := u.Fragment
	hash = strings.TrimPrefix(hash, "!")
	hash = strings.TrimPrefix(hash, "/")
	segs := strings.Split(hash, "/")
	if len(segs) < 3 || segs[0] != "ts" {
		return DetailRef{}, fmt.Errorf("invalid detail URL: %s", raw)
	}
	ft := segs[2]
	if _, err := util.ParseTimestamp(ft); err != nil {
		return DetailRef{}, fmt.Errorf("invalid ft timestamp in detail URL: %s", raw)
	}
	return DetailRef{StationID: segs[1], FT: ft}, nil
}

// PickLatestDetailURL selects the most recent detail URL that is not in the
// future relative to now, which matches expected timeshift availability.
func PickLatestDetailURL(urls []string, now time.Time) (string, error) {
	nowTS := util.FormatTimestamp(now)
	type item struct {
		url string
		ft  string
	}
	items := make([]item, 0, len(urls))
	for _, u := range urls {
		d, err := ExtractDetailFromDetailURL(u)
		if err != nil {
			// Ignore malformed entries and keep scanning other candidates.
			continue
		}
		if d.FT <= nowTS {
			items = append(items, item{url: u, ft: d.FT})
		}
	}
	if len(items) == 0 {
		return "", fmt.Errorf("no usable detail URL found in search results")
	}
	sort.Slice(items, func(i, j int) bool { return items[i].ft > items[j].ft })
	return items[0].url, nil
}
